File: android/joinery-android-member/src/main/java/com/getjoinery/memberkit/MemberApi.kt
package com.getjoinery.memberkit

import com.getjoinery.android.ApiClient
import com.getjoinery.android.JoineryApiError
import com.getjoinery.android.JsonValue

/**
 * Thin typed face over the dashboard/orders/subscriptions/events read actions
 * plus the two reused mutations that back them (`orders_recurring_action`,
 * `event_withdraw`). Every call rides the app's session key through ApiClient;
 * scoping is entirely server-side — same query paths as the web profile pages.
 */
// `open` so a test can subclass and control call timing for the stale-load
// race test; production always uses the concrete methods below.
open class MemberApi(val client: ApiClient) {

    suspend fun dashboard(): DashboardSummary {
        val envelope = client.submitAction("profile_dashboard", JsonValue.Obj(emptyList()))
        return DashboardSummary.from(envelope["data"]) ?: throw JoineryApiError.Malformed
    }

    suspend fun orders(offset: Int): OrderPage {
        val envelope = client.submitAction(
            "order_list",
            JsonValue.obj("offset" to JsonValue.Num(offset.toDouble())),
        )
        return OrderPage.from(envelope["data"]) ?: throw JoineryApiError.Malformed
    }

    suspend fun subscriptions(): SubscriptionSummaryPayload {
        val envelope = client.submitAction("subscription_summary", JsonValue.Obj(emptyList()))
        return SubscriptionSummaryPayload.from(envelope["data"]) ?: throw JoineryApiError.Malformed
    }

    /** Cancel a subscription at period end. The server cancels unconditionally
     *  once called (no server-side confirm step) — the confirmation is the
     *  client's dialog before this call, matching the web page's flow. */
    suspend fun cancelSubscription(orderItemId: Int) {
        client.submitAction(
            "orders_recurring_action",
            JsonValue.obj("order_item_id" to JsonValue.Num(orderItemId.toDouble())),
        )
    }

    open suspend fun events(status: EventStatusFilter, offset: Int): EventPage {
        val envelope = client.submitAction(
            "my_events",
            JsonValue.obj(
                "status" to JsonValue.Str(status.slug),
                "offset" to JsonValue.Num(offset.toDouble()),
            ),
        )
        return EventPage.from(envelope["data"]) ?: throw JoineryApiError.Malformed
    }

    /** Withdraw from an event registration. `confirm: true` matches the web
     *  confirmation form's required field. */
    suspend fun withdraw(registrantId: Int) {
        client.submitAction(
            "event_withdraw",
            JsonValue.obj(
                "evr_event_registrant_id" to JsonValue.Num(registrantId.toDouble()),
                "confirm" to JsonValue.Bool(true),
            ),
        )
    }
}

File: ios/joinery-kit/Sources/JoineryMemberKit/MemberAPI.swift
import Foundation
import JoineryKit

/// Thin typed face over the dashboard/orders/subscriptions/events read
/// actions plus the two reused mutations that back them
/// (`store/orders_recurring_action`, `event_manager/event_withdraw`). Every call rides the
/// app's session key through APIClient; scoping is entirely server-side —
/// same query paths as the web profile pages.
public struct MemberAPI: Sendable {
    let client: APIClient

    public init(client: APIClient) {
        self.client = client
    }

    public func dashboard() async throws -> DashboardSummary {
        let envelope = try await client.submitAction("profile_dashboard", body: .object([]))
        guard let summary = DashboardSummary(data: envelope["data"]) else {
            throw JoineryAPIError.malformedResponse
        }
        return summary
    }

    public func orders(offset: Int) async throws -> OrderPage {
        let envelope = try await client.submitAction("store/order_list", body: .object([
            (key: "offset", value: .number(Double(offset))),
        ]))
        guard let page = OrderPage(data: envelope["data"]) else {
            throw JoineryAPIError.malformedResponse
        }
        return page
    }

    public func subscriptions() async throws -> SubscriptionSummaryPayload {
        let envelope = try await client.submitAction("store/subscription_summary", body: .object([]))
        guard let payload = SubscriptionSummaryPayload(data: envelope["data"]) else {
            throw JoineryAPIError.malformedResponse
        }
        return payload
    }

    /// Cancel a subscription at period end. The server cancels unconditionally
    /// once called (there is no server-side confirm step) — the confirmation
    /// is the client's alert before this call, matching the web page's flow.
    public func cancelSubscription(orderItemID: Int) async throws {
        _ = try await client.submitAction("store/orders_recurring_action", body: .object([
            (key: "order_item_id", value: .number(Double(orderItemID))),
        ]))
    }

    public func events(status: EventStatusFilter, offset: Int) async throws -> EventPage {
        let envelope = try await client.submitAction("event_manager/my_events", body: .object([
            (key: "status", value: .string(status.rawValue)),
            (key: "offset", value: .number(Double(offset))),
        ]))
        guard let page = EventPage(data: envelope["data"]) else {
            throw JoineryAPIError.malformedResponse
        }
        return page
    }

    /// Withdraw from an event registration. `confirm: true` matches the web
    /// confirmation form's required field.
    public func withdraw(registrantID: Int) async throws {
        _ = try await client.submitAction("event_manager/event_withdraw", body: .object([
            (key: "evr_event_registrant_id", value: .number(Double(registrantID))),
            (key: "confirm", value: .bool(true)),
        ]))
    }
}
